### ISSUE TITLE
VS2017 compiler suggests using constexpr if

### ISSUE THREAD

**Issue by DragonOsman** [nlohmann/json#1204]
I'm making a bug report, but it's not really a bug.  It's something between a bug and a feature request.  Basically, the MSVC compiler (latest version in VS2017) is suggesting to use `constexpr if` in the file `single_include\nlohmann\json,hpp` on line 1966.  The warning is: 
```
warning C4127: conditional expression is constant
note: consider using 'if constexpr' statement instead
```
Also, on `single_include\nlohmann\json.hpp(1962)` (`962 is the line number):
```
note: while compiling class template member function 'std::char_traits<char>::int_type nlohmann::detail::wide_string_input_adapter<std::u32string>::get_character(void) noexcept'
```

**Comment by nlohmann** [comment]
Thanks for the info! Unfortunately, I am not sure what to do with it, because we target C++11, and `constexpr if` are a C++17 extension IIRC. So the best I could do to silence this warning would be to decorate it with some macros. Not sure if this is worth it, though...

**Comment by jk-jeon** [comment]
VS2017 now properly support feature testing macros, which others already have supported.
How about this?

````
#ifdef __cpp_if_constexpr
  if constexpr( ... )
#else
  if( ... )
#endif
````

**Comment by theodelrieu** [comment]
I would rather write a metafunction than introduce some ifdefs.

**Comment by nlohmann** [comment]
Fixed by merging #1272.
